design/handle-backup-of-volumes-by-resources-filters.md

```diff
@@ -0,0 +1,246 @@
+- option to skip all PV data
+- option to skip specified PV type (RBD, NFS)
+- option to skip specified PV size
+- option to skip specified storage-class
+ 
+## High-Level Design
+Add a new flag `backup-policies-from-file` when executing `velero backup create`, which imports the defined volume resources policies in one YAML file. the YAML file defined volume resources policies for the current backup.
+ 
+When Velero handles volumes backup should respect the policies defined in the imported YAML file.
+ 
+## Detailed Design
+The volume resources policies should contain a list of policies which is the combination of conditions and related `action`, when target volumes meet the conditions, the related `action` will take effection.
+
+Below is the API Design for the user configuration:
+
+### API Design
+```go
+// Action defined one action for a specific way of backup
+type Action struct {
+	// Type defined specific type of action, it could be 'file-system-backup', 'volume-snapshot', or 'skip' currently
+	Type string `yaml:"type,omitempty"`
+	// Parameters defined map of parameters when executing a specific action
+	// +optional
+	// +nullable
+	Parameters map[string]interface{} `yaml:"parameters,omitempty"`
+}
+
+// VolumePolicy defined policy to conditions to match Volumes and related action to handle matched Volumes
+type VolumePolicy struct {
+	// Conditions defined list of conditions to match Volumes
+	Conditions map[string]interface{} `yaml:"conditions,omitempty"`
+	Action     Action                 `yaml:"action,omitempty"`
+}
+
+// ResourcePolicies currently defined slice of volume policies to handle backup
+type ResourcePolicies struct {
+	VolumePolicies []VolumePolicy `yaml:"volumePolicies,omitempty"`
+	// we may support other resource policies in the future, and they could be added separately
+  	// OtherResourcePolicies: []OtherResourcePolicy
+}
+```
+
+The policies YAML config file would look like this:
+```yaml
+version: v1
+volumePolicies:
+# it's a list and if the input item matches the first policy, the latters will be ignored
+# each policy consists of a list of conditions and an action
+
+# each key in the object is one condition, and one policy will apply to resources that meet ALL conditions
+- conditions:
+    # capacity condition matches the volumes whose capacity falls into the range
+    capacity: "0,100Gi"
+    csi:
+      driver: aws.ebs.csi.driver
+      fsType: ext4
+    storageClass:
+    - gp2
+    - ebs-sc
+  action:
+    type: volume-snapshot
+    parameters:
+      # optional parameters which are custom-defined parameters when doing an action
+      volume-snapshot-timeout: "6h"
+- conditions:
+    capacity: "0,100Gi"
+    storageClass:
+    - gp2
+    - ebs-sc
+  action:
+    type: file-system-backup
+- conditions:
+    nfs: {}
+  action:
+    # type of file-system-backup could be defined a second time
+    type: file-system-backup
+- conditions:
+    csi:
+      driver: aws.efs.csi.driver
```

Comment:
I assume there will be a catch-all condition that we can use at the end? For example if user wants to skip all PVs except storage class abc.

Reply:
whether should we have a `catch-all condition` is depend on user configuration. If users configured it, velero would support it.

The current situation would be the ignored volumes will not be backed up.
suppose we have volumes `A, B, C, and D`, through all kinds of resources filters, A and B are defined to include, D is excluded, Velero will backup volumes `A` and `B`, and `D` won't be backed up. the ignored `C` will not be backed up by default.

Reply:
Yes, I think that's a good enhancement, thanks @rnarenpujari 

cc @qiuming-best

Reply:
> The current situation would be the ignored volumes will not be backed up.
> suppose we have volumes A, B, C, and D, through all kinds of resources filters, A and B are defined to include, D is excluded, Velero will backup volumes A and B, and D won't be backed up. the ignored C will not be backed up by default.  

OK skip was not a good example. If we want to snapshot storage class abc but use restic for everything else, then a catch-all/wildcard condition at the end would be nice to have I think.